src/main/navigation/navigation.c

```diff
@@ -2289,6 +2290,18 @@ void updateClimbRateToAltitudeController(float desiredClimbRate, climbRateToAlti
         posControl.desiredState.pos.z = altitudeToUse;
     }
     else {
+
+        /* 
+         * If max altitude is set, reset climb rate if altitude is reached and climb rate is > 0
+         * In other words, when altitude is reached, allow it only to shrink
+         */
+        if (navConfig()->general.max_altitude > 0 && 
+            altitudeToUse >= navConfig()->general.max_altitude &&
+            desiredClimbRate > 0
+        ) {
+            desiredClimbRate = 0;
```

Comment:
I suggest adding this as a separate controller capping the altitude setpoint itself

Reply:
It's like that on purpose. There are 13 places in the code where `updateClimbRateToAltitudeController` is used. More than that, in some places there is stick->altitude->climbRate (MR) and in some stick->climbRate (FW). I might not be very proud of it, but it's KISS. But agreed, I will refactor this a little

Reply:
If we set max altitude, it should be imposed on all altitudes, including RTH altitude. Which means it should be limiting the `posControl.desiredState.pos.z` itself. 

Limiting the target in `updateClimbRateToAltitudeController` and `setDesiredPosition` should pretty much cover all cases.

Reply:
What is this going to control? motor(s), elevator, spoilers etc?? Please think of altitude limiting for (motor) GLIDERS. Perhaps in Failsafe/RTH situations if the (adustable) max altitude is exceeded then the altitude limiting lasts until a lower (adjustable) altitude is reached.

Reply:
Do you think this could be also applied in manual mode somehow? 
Let me explain: To fly professionally in France, your drone is required to have a programmable max height that cannot be crossed even if you apply full stick control. This is required in S1 to S4 scenarios. Let's stick to S1 (outside populated areas) and S3 (populated areas) as S2 and S4 require boundaries and several other things.

Some pro in France have found a way to modify Betaflight and/or iNav to comply with this ceiling limit, Sadly they are unwilling to share how. As an example, this video shot recently by professionals in S3 with a 5inch quad: https://www.youtube.com/watch?v=90fVvKpuU7Y

Note that, in recreational situations, a simple visual or audio alarm is sufficient. And french authorities have already stated that these S1 to S4 scenarios will remain for at least a few years on top of the European rules.
Also note that contrarily to many believing so, light drones do not need to be certified if they remain below a certain weight limit in S1, S2, S3. It's another story for S4

Reply:
@Spitfire4466  i have always shared my solution , so people with better dev knowledge could improve it , first i ve done it in emu and after beta . feel free to use/improve it